File: 1 Algorithmic Toolbox/Week 1 Programming Challenges/2_maximum_pairwise_product/stress_test.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>

using std::vector;
using std::cin;
using std::cout;
using std::max;

// slow but correct algorithm
long long MaxPairwiseProductNaive(const vector<long long>& numbers) {
    long long max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

long long MaxPairwiseProductFast(const vector<long long>& numbers) {
    int n = numbers.size();
    int index1 = -1;
    int index2 = -1;

    for (int i = 0; i < n; ++i) {
        if ((index1 == -1) || (numbers[i] > numbers[index1])){
            index1 = i;
        }
    }
    
    for (int j = 0; j < n; ++j) {
        if ((j != index1) && ((index2 == -1) || (numbers[j] > numbers[index2]))){
            index2 = j;
        }
    }
    
    //cout << index1 << ' ' << index2;

    return ((long long) (numbers[index1])) * numbers[index2];
}

int main() {
    int N = 10;
    int M = 100000;
    while (true){
        int n = rand() % N + 2;
        cout << n << "\n";
        vector<long long> test;
        for (int i = 0; i < n; ++i){
            test.push_back(rand() % M);
        }
        for (int i = 0; i < n; ++i){
            cout << test[i] << ' ';
        }
        cout << "\n";
        long long res1 = MaxPairwiseProductNaive(test);
        long long res2 = MaxPairwiseProductFast(test);
        if (res1 != res2){
            cout << "Wrong Answer: " << res1 << ' ' << res2 << "\n";
            break;
        }
        else{
            cout << "OK\n";
        }
    }
}

File: 1 Algorithmic Toolbox/Week 1 Programming Challenges/2_maximum_pairwise_product/stress_test.py
# -*- coding: utf-8 -*-
"""
Created on Wed Feb 13 11:25:44 2019

@author: Yu Chen
"""

# python3
import random

# slow but correct algorithm
def max_pairwise_product_naive(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                              numbers[first] * numbers[second])

    return max_product

def max_pairwise_product_fast(numbers):
    n = len(numbers)
    index1 = -1
    for i in range(n):
        if ((index1 == -1) | (numbers[i] > numbers[index1])):
            index1 = i
    index2 = -1
    for j in range(n):
        if ((j != index1) & ((index2 == -1) | (numbers[j] > numbers[index2]))):
            index2 = j

    return numbers[index1] * numbers[index2]

if __name__ == '__main__':
    N = 10
    M = 100000
    while(True):
        num = random.randint(2, N)
        print("\n")
        test = []
        for i in range(num):
            test.append(random.randit(M))
        for i in range(num):
            print(test[i], end=" ")
        print("\n")
        res1 = max_pairwise_product_naive(test)
        res2 = max_pairwise_product_fast(test)
        if (res1 != res2):
            print("Wrong Answer: ", res1, " ", res2, "\n")
        else:
            print("OK\n")
    
